src/Test/TestCases.Runtime/ParallelBranchTests.cs

```diff
@@ -0,0 +1,157 @@
+{
+    protected override IWorkflowSerializer Serializer => new JsonWorkflowSerializer();
+}
+
+public class ParallelBranchTests
+{
+    public class TestCodeActivity(Action<CodeActivityContext> onExecute) : CodeActivity
+    {
+        protected override void Execute(CodeActivityContext context)
+        {
+            onExecute(context);
+        }
+    }
+
+    protected virtual IWorkflowSerializer Serializer => new DataContractWorkflowSerializer();
+    string _parentLevel = default;
+    readonly string _noBranch = default;
+
+    private void Run(params Action<CodeActivityContext>[] onExecute)
+    {
+        new WorkflowApplication(new SuspendingWrapper(onExecute.Select(c => new TestCodeActivity(c))))
+        {
+            InstanceStore = new MemoryInstanceStore(Serializer)
+        }.RunUntilCompletion();
+    }
+
+    private void RunWithParent(params Action<CodeActivityContext>[] onExecute)
+    {
+        Run([SetParent, .. onExecute]);
+        void SetParent(CodeActivityContext context)
+        {
+            var parent = context.CurrentInstance.Parent;
+            parent.GetCurrentParallelBranchId().ShouldBe(_noBranch);
+            parent.MarkNewParallelBranch();
+            _parentLevel = parent.GetCurrentParallelBranchId();
+        }
+    }
+
+    [Fact]
+    public void Push()
+    {
+        var level1 = ParallelTrackingExtensions.PushNewBranch(null);
+        var level2 = ParallelTrackingExtensions.PushNewBranch(level1);
+        var level3 = ParallelTrackingExtensions.PushNewBranch(level2);
+        level2.ShouldStartWith(level1);
+        level3.ShouldStartWith(level2);
+        var l3Splits = level3.Split('.');
+        l3Splits.Length.ShouldBe(3);
+        l3Splits.First().ShouldBe(level1);
+    }
+
+    [Fact]
+    public void SetToNullWhenNull() => Run(
+    context =>
+    {
+        context.CurrentInstance.SetCurrentParallelBranchId(_noBranch);
+        context.CurrentInstance.SetCurrentParallelBranchId(_noBranch);
+        context.CurrentInstance.GetCurrentParallelBranchId().ShouldBeNull();
+    });
+
+    [Fact]
+    public void SetToNullWhenNotNull() => Run(
+    context =>
+    {   
+        context.CurrentInstance.SetCurrentParallelBranchId(_noBranch);
+        context.CurrentInstance.SetCurrentParallelBranchId(ParallelTrackingExtensions.PushNewBranch(_noBranch));
+        context.CurrentInstance.SetCurrentParallelBranchId(_noBranch);
+        context.CurrentInstance.GetCurrentParallelBranchId().ShouldBeNull();
+    });
+
+
+    [Fact]
+    public void GetCurrentParallelBranch_InheritsFromParent() => RunWithParent(
+    context =>
+    {
+        var branchId = context.GetCurrentParallelBranchId();
+        var currentBranch = context.CurrentInstance.GetCurrentParallelBranchId();
+        currentBranch.ShouldBe(branchId);
+        currentBranch.ShouldBe(_parentLevel);
+    });
+
+    [Fact]
+    public void PushAndSetParallelBranch() => RunWithParent(
```

Comment:
I think the "Push" naming is not helping at all - hopefully we remove it.
We are basically validating that the getter returns what was set, so I'd name it SetAndGetBranchId